src/Features/ExternalAccess/Copilot/Internal/SemanticSearch/CopilotSemanticSearchQueryExecutor.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.ExternalAccess.Copilot.SemanticSearch;
+using Microsoft.CodeAnalysis.SemanticSearch;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.FindUsages;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Notification;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Classification;
+using Roslyn.Utilities;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Copilot.Internal.SemanticSearch;
+
+[Export(typeof(ICopilotSemanticSearchQueryExecutor)), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class CopilotSemanticSearchQueryExecutor(IHostWorkspaceProvider workspaceProvider) : ICopilotSemanticSearchQueryExecutor
+{
+    private sealed class ResultsObserver(CancellationTokenSource cancellationSource, int resultCountLimit) : ISemanticSearchResultsObserver
+    {
+        public List<string> Results { get; } = [];
+        public string? RuntimeException { get; private set; }
+        public bool LimitReached { get; private set; }
```

Comment:
can this just be computed, instead of stored as a field?

Reply:
`Results.Count == limit` might mean there are actually exactly `limit` items or there is more and we stopped populating them.